### DISCUSSION TITLE
Z offset problem

### DISCUSSION THREAD

**Discussion by DomanRulez** [xblax/flashforge_ad5m_klipper_mod#61] [Q&A]
Hi
My Z offset was too large, so I'm trying to calibrate it using the console. When I try to save the settings and restart Klipper, I get this error:
**SAVE_CONFIG section, 'probe' option 'z_offset' conflicts with attached value**
I'm also trying to do this using Klipper's screen, but I have the same error
![image](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/161459652/5a270eb0-1157-4035-a336-e15d81de6994)

settings in printer.base.cfg:
![image](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/161459652/d9bd4554-82ab-4bda-95f7-9efc6bdd7a4c)

**Comment by isod89** [comment]
I don't know if it's something to do with the fact that I'm a beginner with Klipper, I'm having the same problem as you.

**Comment by consp** [comment]
I never actually adjusted the z-offset. Simple bed calibration was enough with the default of -0.25. Not my field of expertise though.

**Comment by xblax** [comment]
I think it's problem with the config. The z offset must be moved into printer.cfg from printer.base.cfg to allow calibration.
But calibrated z-offset of -0.5 looks way to much, though. -0.25 is the default from Flashforge, and for me it's more or less spot-on.

**Comment by xblax** [comment]
@DomanRulez @isod89 
This should fix it. You can edit your config until this will be in the next release:
https://github.com/xblax/flashforge_adm5_klipper_mod/commit/8309829b0a69ce625ac81950660cf6bd2d7669b9

**Reply by DomanRulez** [reply]
Thank you, I will check this solution when I get back from work

**Comment by Tiwatz** [comment]
I tried this in Mainsail. If I adjust the Z-offset on dashboard that seems to offset the z-position endstop when I hit Save. So to fix that we need to move that line on to printer.cfg

[stepper_z]
position_endstop: 220

![Z](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/162510297/d1428861-6a15-4230-87c3-31e865ab83f0)

**Reply by xblax** [reply]
I don't think that's the correct approach, because we don't use the probe as end stop.
You should rather use `PROBE_CALIBRATE`, which will just adjust the z-offset.

**Reply by xblax** [reply]
Or you can use `Z_OFFSET_APPLY_PROBE`, that works as well. But the endstop should not be changed.

**Reply by Tiwatz** [reply]
Agree. But that's seems to be the way Mainsail does it when I use the z-offset panel in the UI. I don't know if it is possible  to change what the save button do?

**Comment by consp** [comment]
The issue also happens with z-offset calibration in klipperscreen. My guess is any changes in the normal version will also fix that. My record is 5mm offset now.

**Comment by xblax** [comment]
It seems to be issue of Klipper (or the GUI Frontends ) how they handle Z_OFFSET_APPLY_ENDSTOP and Z_OFFSET_APPLY_PROBE.
Mainsail does "auto-detection" but selects the wrong one by default. It can be overridden in the menu: 

![image](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/9802544/d4861c22-2249-483c-8276-c35c40181157)

Fluidd had a direct menu. It's a bit buggy, because when changing the option twice it want's to apply the offset to both ENDSTOP and PROBE.

![image](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/9802544/342c9ba5-39ae-42c2-b54e-fc5b79a413f4)

Unfortunately Z_OFFSET_APPLY_ENDSTOP also cannot be disabled in Klipper.

Saving offset to endstop makes absolutely not sense for. Because after probing, distance from bed will be the same as before. It must be corrected again and then the endstop position will get offset (in an endless loop).

Should we override Z_OFFSET_APPLY_ENDSTOP and shot a warning to use Z_OFFSET_APPLY_PROBE instead, or should we just always do Z_OFFSET_APPLY_PROBE even for the Z_OFFSET_APPLY_ENDSTOP command?

**Comment by DougJoseph** [comment]
I benefitted from this discussion, and edited to move the offset from "base" cfg to printer.cfg, and also changed the Mainsail GUI option to apply to probe instead of end stop.

I do have one question about using the Klipperscreen GUI for doing probe option to tweak the Z offset:

I'm pretty brand new to Klipper, but not new to 3D printing. 

On the Klipperscreen GUI option for probing for offset, when it comes to the part where I manually adjust, am I supposed to be using the old "sheet of paper between the nozzle and bed" trick?

**Reply by xblax** [reply]
> On the Klipperscreen GUI option for probing for offset, when it comes to the part where I manually adjust, am I supposed to be using the old "sheet of paper between the nozzle and bed" trick?

Yes that's the best option. See https://www.klipper3d.org/Probe_Calibrate.html#calibrating-probe-z-offset
The general idea is that nozzle presses against the bed during probing but it moves a bit further on the z-axis than needed for light contact, because the probe only triggers when the force onto the bed becomes high enough. Therefore after leveling, the nozzle must be offset by a fixed distance to be directly above the bed.

**Reply by DougJoseph** [reply]
Thanks!!!
